numba_mlir/numba_mlir/mlir_compiler/lib/pipelines/PlierToLinalg.cpp

```diff
@@ -4109,6 +4109,143 @@ static void populateCommonOptPass(mlir::OpPassManager &pm) {
       }));
 }
 
+struct RemoveExtraCopyForOut: public mlir::OpRewritePattern<mlir::func::CallOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::func::CallOp op,
+                  mlir::PatternRewriter &rewriter) const override {
+    using mlir::linalg::GenericOp;
+    using mlir::linalg::YieldOp;
+    using mlir::dyn_cast_or_null;
+    using mlir::memref::AllocOp;
+    using mlir::memref::AllocaOp;
+
+    if (op.getOperands().empty())
+      return mlir::failure();
+
+    auto hasIdentityMaps = [](GenericOp& Op) {
+      auto maps = Op.getIndexingMaps();
+      if (maps.size() != 2)
+        return false;
+
+      bool trivialAffineMaps = true;
+      for (auto&& map : maps)
+        trivialAffineMaps &= map.cast<mlir::AffineMapAttr>().getValue().isIdentity();
+
+      return trivialAffineMaps;
+    };
+
+    auto hasCopyBody = [](GenericOp& Op) {
+      auto* body = Op.getBody();
+      if (llvm::hasSingleElement(*body)) {
+        if (auto yield = dyn_cast_or_null<YieldOp>(body->front())) {
+          auto operands = yield.getOperands();
+          if (operands[0] == body->getArgument(0))
+            return true;
+        }
+      }
+
+      return false;
+    };
+
+    auto hasOnlyCopyUser = [hasIdentityMaps, hasCopyBody](mlir::Value& operand) -> GenericOp {
+      if (!llvm::hasNItems(operand.getUsers(), 2))
+        return {};
+
+      for (auto&& user : operand.getUsers()) {
+        if (auto linGen = dyn_cast_or_null<GenericOp>(*user)) {
+          if (hasIdentityMaps(linGen) && hasCopyBody(linGen))
+            return linGen;
+        }
+      }
+
+      return {};
+    };
+
+    auto potentiallyDominates = [](auto op, auto another) {
+      auto dinfo = mlir::DominanceInfo();
+
+      if (dinfo.properlyDominates(op, another))
+        return true;
+
+      if (dinfo.properlyDominates(another, op)) {
+        bool operands_dominates = true;
+        for (auto&& oper : op->getOperands())
+          operands_dominates &= dinfo.properlyDominates(oper, another);
+
+        return operands_dominates;
+      }
+
+      return false;
+    };
+
+    llvm::SmallVector<mlir::Operation*> OpsToMove;
+    llvm::SmallVector<std::tuple<mlir::Value, mlir::Value, GenericOp>> ValsToReplace;
+
+    for (auto&& operand : op.getOperands()) {
+      auto opDefOp = operand.getDefiningOp();
+      if (!mlir::isa_and_nonnull<AllocOp, AllocaOp>(opDefOp))
+        continue;
+
+      auto linGen = hasOnlyCopyUser(operand);
```

Comment:
You will need to also check if there are no side-effecting ops between call and copy op.